Fix invalid casts int and float

kkmuffme opened the pull request:
e.g. https://psalm.dev/r/0d284c6f48 gives `Warning: Object of class stdClass could not be converted to int`

Since by default there cannot be any "PossiblyInvalidCasts", as we always get 0/0.0 in error case, added mostly unwanted typecasts that are a leading cause of off-by-one errors as PossiblyInvalidCasts

Fix: https://github.com/vimeo/psalm/issues/6966 (the remaining half + cherry-pick the other half from master to 4.x)

kkmuffme commented:
Feedback appreciated :-)

Could someone please help with the failing taint tests?

AndrolGenhald commented on tests/ArrayFunctionCallTest.php:
```diff
@@ -2163,7 +2163,7 @@ function($val) { return (int)$val; },
                 ',
                 'assertions' => [
-                    '$line===' => 'array{0: int}<array-key, int>',
+                    '$line===' => 'array{0: 0}<array-key, 0>',
```
This seems wrong. What if `$line` was `["123", "456"]`?

kkmuffme replied:
Then the inferred type will be correctly set to `array{0: 123, 1: 456}`. But the given array is `["a", "b"]` which, if you typecast will be 0 for both values.

AndrolGenhald replied:
But shouldn't `$line` be `array{0: string}<string>` when `array_map` is called? At the point the callback is evaluated it should only know that `$val` is a `string` shouldn't it?

AndrolGenhald commented:
There's a lot going on here that I don't quite understand, could you add a description of the issue you're trying to solve?

kkmuffme commented:
>There's a lot going on here that I don't quite understand, could you add a description of the issue you're trying to solve?

Summary: this PR adds correct support for typecasting to int/float. 

- typecasting of objects/resources to int/float gives a PHP warning, but Psalm doesn't say anything about that. This PR adds InvalidCast error for this, https://psalm.dev/r/0d284c6f48

- typecasting arrays to int/float is always 0 or 1, not generic int https://psalm.dev/r/9e15ff2d1f

- due to this 0/1, typecasting of arrays often leads to off-by-one errors  in code, that are extremely hard to track down in some cases. This is now added as "PossiblyInvalidCast" for int/float casts of arrays.

- typecasting (string) true is literal string '1', not generic string.

>If taint tests (or unused variable tests) are failing, you probably messed up $parent_nodes somewhere.

Yes, thanks. That was it, the first commit that just unified the code messed it up. Reverted it now (will make it look nice later).
Now there are some other errors, e.g. for (array) casts, even though they weren't changed.
Some input on the failing tests would be great.

AndrolGenhald commented:
> Now there are some other errors, e.g. for (array) casts, even though they weren't changed.
> Some input on the failing tests would be great.

I think the array one that's failing is the one I mentioned previously, looks like the rest of them are some taint tests that are still failing.

orklah commented on src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php:
```diff
@@ -268,7 +254,7 @@ public static function analyze(
             }
 
-            if ($statements_analyzer->data_flow_graph) {
+            if ($statements_analyzer->data_flow_graph instanceof VariableUseGraph) {
```
does reverting this change helps with taint errors?

AndrolGenhald commented:
> This is now added as "PossiblyInvalidCast" for int/float casts of arrays.

@orklah What do you think about adding `DangerousCast` for this instead? I think `PossiblyInvalidCast` should be used when casting a variable whose type is a union where the cast might be valid or might be invalid (although I think there are probably some issues where we use a `PossiblyInvalid...` issue when it should just be `Invalid...` and vice versa).

orklah commented:
Yeah, that makes sense, we should try to keep this rule consistent

AndrolGenhald commented on src/Psalm/Issue/RiskyCast.php:
```diff
@@ -0,0 +1,9 @@
+class RiskyCast extends CodeIssue
+{
+    public const ERROR_LEVEL = 3;
+    public const SHORTCODE = 303;
```
@orklah We need to think about conflicting short codes with Psalm 5...

AndrolGenhald commented on src/Psalm/Internal/Cli/Refactor.php:
```diff
@@ -284,7 +284,7 @@ function () use ($current_dir, $options, $vendor_dir): ?\Composer\Autoload\Class
         }
 
-        $threads = isset($options['threads'])
+        $threads = isset($options['threads']) && is_numeric($options['threads'])
```
Should we add some sort of notice if it's non-numeric?

AndrolGenhald replied:
> However psalm's behavior has not changed

Actually it has a bit: https://3v4l.org/Xj0BS, don't you just love the way PHP handles casts?

I think this is probably fine though.

AndrolGenhald commented on src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php:
```diff
@@ -352,7 +778,14 @@ public static function castStringAttempt(
+            }
+
             if ($atomic_type instanceof TMixed
+                || $atomic_type instanceof TClosedResource
```
Does PHP ever give a warning for this? Do we need to add an issue in some cases?

AndrolGenhald replied:
According to [this](https://www.php.net/manual/en/language.types.string.php#language.types.string.casting):

> Resources are always converted to strings with the structure "Resource id #​1", where 1 is the resource number assigned to the resource by PHP at runtime. While the exact structure of this string should not be relied on and is subject to change, it will always be unique for a given resource within the lifetime of a script being executed (ie a Web request or CLI process) and won't be reused.

so this could probably be non-empty-string.

AndrolGenhald replied:
> And resource/closed-resource TIntRange(1, null) for typecasts to (int) too

I see [this](https://www.php.net/manual/en/language.types.integer.php#language.types.integer.casting) which says:

> If a resource is converted to an int, then the result will be the unique resource number assigned to the resource by PHP at runtime.

I don't see anything guaranteeing it's positive, and I found [this](https://stackoverflow.com/a/6946116) saying non-positive numbers aren't expected, but they are possible. @orklah what do you think about that? Maybe `int` is safer here since I'm not sure if a non-positive value is considered a bug by PHP?

AndrolGenhald commented on src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php:
```diff
@@ -300,6 +289,444 @@ public static function analyze(
+                continue;
+            }
+
+            if ($atomic_type instanceof TBool) {
+                $valid_ints[] = new TLiteralInt(1);
+                continue;
+            }
```
Either you meant `TTrue` instead of `TBool`, or this should be `int<0, 1>`. Probably handle `TTrue` first and then `TBool`?

AndrolGenhald replied:
> No, this is correct. TFalse is already handled in the condition above (TLiteralInt(0)) therefore bool is only TTrue here.

Actually no. The type `false` is represented by `TFalse`, `true` is represented by `TTrue`, but bool isn't `Union([new TFalse(), new TTrue()])`, it's `TBool`, so you can absolutely have a `TBool` that's neither `TTrue` nor `TFalse`.

AndrolGenhald commented on src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php:
```diff
@@ -300,6 +289,444 @@ public static function analyze(
+            }
+
+            // could be invalid, but allow it, as it is allowed for TString below too
+            if ($atomic_type instanceof TMixed
```
@orklah What do you think about [this behavior](https://psalm.dev/r/170b48f07d)? Leave it as is to avoid a bunch of `mixed` issues?

It annoys me a bit because I sometimes have to use `mixed` to represent an unknown value that's able to be anything, but I still want to be warned if I do anything invalid with it. Unfortunately for me Psalm likes to ignore a lot of problems with `mixed` types because otherwise untyped codebases would have a ridiculous number of errors.

orklah replied:
Well, in my mind at least, whenever I see an explicit cast, I tend to assume the developper knew what he was doing. I don't know if this hunch is correct and verified in the wild, but I would be personnally annoyed if Psalm were to report me such cases (If the projects I analyze were at level 1 at least!).

AndrolGenhald replied:
I never like to assume the developer knows what they're doing! :stuck_out_tongue_closed_eyes:

But yeah, it fits with all the other special exceptions we have for `mixed`. Maybe I'll have to start using `object|resource|array|string|int|float|bool|null` instead, despite the verbosity.

orklah replied:
I was going to say that Psalm would revert that to mixed [like we do with scalar](https://github.com/vimeo/psalm/blob/b4fdc3e326fea865de49827f723a2703bb18617a/src/Psalm/Internal/Type/TypeCombiner.php#L334) but it doesn't seem so!

AndrolGenhald commented on src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php:
```diff
@@ -300,6 +289,444 @@ public static function analyze(
+                    ) {
+                        $castable_types[] = new TInt();
+
+                        continue 2;
```
Same as above.

AndrolGenhald commented on src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php:
```diff
@@ -300,6 +289,444 @@ public static function analyze(
+                continue;
+            }
+
+            // always 1 for "error" cases
+            $valid_ints[] = new TLiteralInt(1);
```
Why `1` and not `int`?

AndrolGenhald replied:
None that I can think of, but I don't have a lot of time right now. @orklah might want to do a quick double check before merging.

AndrolGenhald commented on src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php:
```diff
@@ -300,6 +289,444 @@ public static function analyze(
+                $statements_analyzer->getSuppressedIssues()
+            );
+        } elseif ($explicit_cast && !$castable_types) {
+            // todo: emit error here
```
Should this be `InvalidCast`?

AndrolGenhald replied:
> But this can never be true, therefore I think we can remove this condition?

Probably, but if you like that can wait for #8486. It's probably easier to refactor these separate casts if they're currently as similar as possible.

AndrolGenhald commented on src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php:
```diff
@@ -300,6 +289,444 @@ public static function analyze(
+
+        $risky_cast = [];
+        $invalid_casts = [];
+        $valid_ints = [];
+        $castable_types = [];
```
Not sure why there is both `$valid_ints` and `$castable_types`, is it just organizational because `$valid_ints` only holds literals? Mind adding a short comment?

AndrolGenhald commented:
> @AndrolGenhald rebased and issues fixed, except "ComplexMethod" in these 2 functions as it doesn't make sense to rework this now, when this has to be modified again in #8486 and [#8366 (review)](https://github.com/vimeo/psalm/pull/8366#pullrequestreview-1107780547)
> 
> I think fixing these 2 issues now, will just make it harder for the followup work.

Agreed, could you `@psalm-suppress` them though so they don't break other PRs once merged?

Also, it looks like there's now a regression in the bitwiseAssignment test.

AndrolGenhald commented on src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php:
```diff
@@ -300,6 +291,457 @@ public static function analyze(
+                ),
+                $statements_analyzer->getSuppressedIssues()
+            );
+        } elseif (!empty($risky_cast)) {
```
@orklah didn't we talk about banning `empty` in the Psalm config at some point? It does seem like a rather pointless function since it basically just casts everything to `bool`. Personally I prefer an explicit `count($arr) !== 0` or `$arr !== []`, but we should probably have some sort of consistency.

orklah replied:
We still have a few to remove, but yeah, best not to add new ones

AndrolGenhald replied:
I don't see it, did you forget to push that part?

AndrolGenhald commented on src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php:
```diff
@@ -300,6 +291,457 @@ public static function analyze(
+                        $intersection_method_id = new MethodIdentifier(
+                            $intersection_type->value,
+                            '__tostring'
+                        );
```
Of course `DOMDocument` causes issues... Are you sure it uses `__toString` though? I sort of figured `DOMDocument` had its own weird cast handling? Maybe we just have to add a special case where it's a `TNamedObject` with `strtolower($...->value) === "domdocument"` (and maybe `"domnode"`, or maybe we should check if it extends `"domnode"`)?

@orklah do you have any idea how `DOM` casting works? I've never looked into it.

AndrolGenhald replied:
I think the best way forward at this point is to just pretend all casts from objects extending `SimpleXMLElement`, `DOMNode`, `GMP`, or `Decimal\Decimal` are valid. We can try to get more specific later if it's possible.

AndrolGenhald replied:
Try `$statements_analyzer->getCodebase()->classExtends(...)`, should also make the check a lot easier compared to what you have now. Also, maybe call it `$pseudo_castable_classes` since it's not really related to `__toString`, and (just to be _really_ nitpicky) make it a private const instead of a static var since it shouldn't change.

AndrolGenhald commented:
Out of curiosity, what was the issue with the broken bitwiseAssignment test?

AndrolGenhald commented:
Huh, weird. Guess there's other bugs around causing that then. @orklah everything look good to you?

orklah commented:
Code seems good, could you add some tests for RiskyCast please?

orklah commented:
Thanks!
